src/main/cpp/tests/cast_string.cpp

```diff
@@ -229,4 +231,212 @@ TYPED_TEST(StringToIntegerTests, Overflow)
+
+    CUDF_TEST_EXPECT_COLUMNS_EQUAL(result->view(), expected);
+  }
+
+  {
+    auto const str = test::strings_column_wrapper({"10.0"});
+    strings_column_view scv{str};
+
+    auto const result =
+      spark_rapids_jni::string_to_decimal(3, 1, scv, false, rmm::cuda_stream_default);
+    test::fixed_point_column_wrapper<int32_t> expected({100}, {1}, numeric::scale_type{-1});
+
+    test::print(result->view());
+
+    CUDF_TEST_EXPECT_COLUMNS_EQUAL(result->view(), expected);
+  }
+
+  {
+    auto const str = test::strings_column_wrapper({"1.7142857343"});
+    strings_column_view scv{str};
+
+    auto const result =
+      spark_rapids_jni::string_to_decimal(9, 8, scv, false, rmm::cuda_stream_default);
+    test::fixed_point_column_wrapper<int32_t> expected({171428573}, {1}, numeric::scale_type{-8});
+
+    test::print(result->view());
+
+    CUDF_TEST_EXPECT_COLUMNS_EQUAL(result->view(), expected);
+  }
+
+  {
+    auto const str = test::strings_column_wrapper({"1.71428573437482136712623"});
+    strings_column_view scv{str};
+
+    auto const result =
+      spark_rapids_jni::string_to_decimal(9, 8, scv, false, rmm::cuda_stream_default);
+    test::fixed_point_column_wrapper<int32_t> expected({171428573}, {1}, numeric::scale_type{-8});
+
+    test::print(result->view());
+    CUDF_TEST_EXPECT_COLUMNS_EQUAL(result->view(), expected);
+
+    auto const fail_result =
+      spark_rapids_jni::string_to_decimal(9, 9, scv, false, rmm::cuda_stream_default);
+    test::fixed_point_column_wrapper<int32_t> null_col({0}, {0}, numeric::scale_type{-9});
+
+    test::print(fail_result->view());
+    CUDF_TEST_EXPECT_COLUMNS_EQUAL(fail_result->view(), null_col);
+  }
+
+  {
+    auto const str = test::strings_column_wrapper({"12.345678901"});
+    strings_column_view scv{str};
+
+    auto const result =
+      spark_rapids_jni::string_to_decimal(9, 8, scv, false, rmm::cuda_stream_default);
+    test::fixed_point_column_wrapper<int32_t> expected({0}, {0}, numeric::scale_type{-8});
+
+    test::print(result->view());
+
+    CUDF_TEST_EXPECT_COLUMNS_EQUAL(result->view(), expected);
+  }
+
+  {
+    auto const str = test::strings_column_wrapper({"0.12345678901"});
+    strings_column_view scv{str};
+
+    auto const result =
+      spark_rapids_jni::string_to_decimal(6, 6, scv, false, rmm::cuda_stream_default);
+    test::fixed_point_column_wrapper<int32_t> expected({123457}, {1}, numeric::scale_type{-6});
+
+    test::print(result->view());
+
+    CUDF_TEST_EXPECT_COLUMNS_EQUAL(result->view(), expected);
+  }
+
+  {
+    auto const str = test::strings_column_wrapper({"1.2345678901"});
+    strings_column_view scv{str};
+
+    auto const result =
+      spark_rapids_jni::string_to_decimal(6, 6, scv, false, rmm::cuda_stream_default);
+    test::fixed_point_column_wrapper<int32_t> expected({0}, {0}, numeric::scale_type{-6});
+
+    test::print(result->view());
+
+    CUDF_TEST_EXPECT_COLUMNS_EQUAL(result->view(), expected);
+  }
```

Comment:
Does this support values with a negative java/scala/spark scale?  If so we are missing tests for this. If not then we should document it.  Support for it is turned off by default in Spark, so I don't think it is a big deal for us to not support it right now.

Reply:
I believe it will, but I haven't tested it. I need to do that and then decide about fixing or adding comments about it. If this doesn't support negative scale, I would need to adjust the plugin to still be able to call the original code in those cases, right?

Reply:
Yes

Reply:
I added a test for negative scales, but I'm not 100% these match spark. I tried to enable negative scales, but it still complained. Need to figure this out so I can verify negative scales are working as intended.
```
scala> spark.conf.set("spark.sql.legacy.allowNegativeScaleOfDecimal", true)

scala> spark.sql("""select cast("1.234e-1" as decimal(6,-2))""").show
org.apache.spark.sql.catalyst.parser.ParseException:
extraneous input '-' expecting INTEGER_VALUE(line 1, pos 36)

== SQL ==
select cast("1.234e-1" as decimal(6,-2))
------------------------------------^^^

  at org.apache.spark.sql.catalyst.parser.ParseException.withCommand(ParseDriver.scala:266)
  at org.apache.spark.sql.catalyst.parser.AbstractSqlParser.parse(ParseDriver.scala:127)
  at org.apache.spark.sql.execution.SparkSqlParser.parse(SparkSqlParser.scala:51)
  at org.apache.spark.sql.catalyst.parser.AbstractSqlParser.parsePlan(ParseDriver.scala:77)
  at org.apache.spark.sql.SparkSession.$anonfun$sql$2(SparkSession.scala:616)
  at org.apache.spark.sql.catalyst.QueryPlanningTracker.measurePhase(QueryPlanningTracker.scala:111)
  at org.apache.spark.sql.SparkSession.$anonfun$sql$1(SparkSession.scala:616)
  at org.apache.spark.sql.SparkSession.withActive(SparkSession.scala:775)
  at org.apache.spark.sql.SparkSession.sql(SparkSession.scala:613)
  ... 47 elided
```

Reply:
Finished testing positive scale values thanks to Bobby's help:
```
import org.apache.spark.sql.types._
spark.conf.set("spark.sql.legacy.allowNegativeScaleOfDecimal", "true")
Seq("-7.691546687437984E16").toDF().withColumn("dec", col("value").cast(DecimalType(10, -1))).show()
```